### PR TITLE
fix(core): Improve model sub-nodes error handling

### PR DESCRIPTION
**Pull request by burivuhster**
### Problem
The current error handling in LLM sub-nodes has two main issues:
1. Model sub-nodes sometimes appear as "running" even after failing
2. n8n incorrectly indicate that errors originated in the root node, when they actually occurred in Model sub-nodes

### Context
The error handling relies on custom exception classes (e.g., `NodeOperationError`). However, there's no straightforward mechanism to wrap model exceptions before they propagate to the root node.

### Solution Approaches Explored
1. Using `N8nLlmTracing` (callback handler passed to the LangChain model's contructor). Using `handleLLMError` callback handler allows to add proper error info to the sub-node and fix the infinite running issue. But it can not be used for wrapping the exception for the root node.
2. `onFailedAttempt` callback can be provided during each model's LangChain class instantiation. It allows to intercept the retry logic built into LangChain, and re-throw the wrapped exceptions.

### Implementation
To use the second approach, the new helper function `makeN8nLlmFailedAttemptHandler` was implemented. It catches LLM errors, wrap them with our custom error classes, provides some default error handling logic (interrupt running if it does not make sense to retry for the specific error), and also allows to customize error handling for the specific cases (for example to use a custom error message for the OpenAI's insufficient quota error).

### Known issues
Couple of LangChain model classes do not use the `AsyncCaller` for their calls. It means, that the common retry logic built into LangChain does not work in such cases. The `onFailedAttempt` callback is not being called as well.
Thus, the described error wrapping logic does not work yet for the following models:
* `LMChatOllama`
* `LmChatAwsBedrock`
* `LmChatGoogleVertex`

Example 1. Insufficient quota error (OpenAI)
Before:
image

After:
image

Example 2. Empty prompt (Anthropic)
Before:
image

After:
image

Example 3. Nested sub-node error
Before:
image

After:
image

Example 4. Nested sub-node error
Before:
image

After:
image

## Related Linear tickets, Github issues, and Community forum posts

- https://linear.app/n8n/issue/AI-298/errors-happen-in-the-root-node-not-in-the-sub-node-even-if-it-belongs
- https://github.com/n8n-io/test-workflows/pull/280

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

**Comment by jeanpaul** [comment]
One thing I noticed, that's a difference now is that I don't really get a sense for what's wrong with my credential, whereas on master I do:

image

Why did that change?

**Comment by jeanpaul** [comment]
Also, the error toast still says "Problem in node 'Basic LLM Chain'", not "OpenAI Chat Model"

**Comment by jeanpaul** [comment]
> One thing I noticed, that's a difference now is that I don't really get a sense for what's wrong with my credential, whereas on master I do:
> 
> image
> Now it just says something is wrong with my credentials:
> 
> image
> Why did that change? I think getting the things to check directly in the toast is a better experience.

I don't know why, but now I cannot reproduce this anymore.

File: packages/editor-ui/src/composables/usePushConnection.ts
Comment (jeanpaul) [inline packages/editor-ui/src/composables/usePushConnection.ts:1]: On line 326 is another occurrence of `context.functionality` -- can you check if that needs to change too?

Comment (jeanpaul) [reply packages/editor-ui/src/composables/usePushConnection.ts:1]: And same for `packages/editor-ui/src/utils/expressions.ts` line 63.

File: packages/@n8n/nodes-langchain/nodes/llms/N8nLlmTracing.ts [hunk 1]
```diff
@@ -175,6 +181,8 @@ export class N8nLlmTracing extends BaseCallbackHandler {
 		runId: string,
 		parentRunId?: string | undefined,
 	) {
+		const runDetails = this.runsMap[runId] ?? { index: Object.keys(this.runsMap).length };
```
Comment (jeanpaul) [inline packages/@n8n/nodes-langchain/nodes/llms/N8nLlmTracing.ts:184]: why do you take the number of keys as the index if `runId` is not in `runsMap`?

**Review by jeanpaul — CHANGES_REQUESTED** [review]

File: packages/editor-ui/src/composables/usePushConnection.ts
Comment (burivuhster) [reply packages/editor-ui/src/composables/usePushConnection.ts:1]: Should be fine as this is for another case, when nodes use "pairedItem" feature (e.g. referencing previous node's item value when the pairedItem connection was interrupted)

Comment (burivuhster) [reply packages/editor-ui/src/composables/usePushConnection.ts:1]: It's being set in `n8nLlmFailedAttemptHandler.ts` when we wrap the error into NodeApiError. For the paired item exceptions – nothing changes, I guess it's being set in `WorkflowDataProxy.ts` and should not be affected by this PR.

Comment (burivuhster) [reply packages/editor-ui/src/composables/usePushConnection.ts:1]: Ahh, I see your point now. I found that setting `context.functionality` to `configuration-node` on NodeApiError does nothing, as it checks for the functionality on the exception object itself. Now I see that the checks are different for the `configuration-node` functionality and `pairedItem`. Good catch, thank you!

**Review by jeanpaul — APPROVED** [review]
Looks good! thanks for addressing the issues with `context.functionality`, and for adding the extra tests! 👍

**Comment by janober** [comment]
Got released with [`n8n@1.68.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.68.0)

**Pull request merged**